shapeshift/web
# feat: yearn opportunities full multi-account support

## Description

- Overview
  - [x] Account selection
  - [x] Use selected account data
- [x] Deposit
  - [x] Approve
  - [x] Confirm
  - [x] Use selected account data in all steps
  - [ ] ~~Poll (regression test against account 0 on prod)~~ WS is super borked atm and Txs messages are not incoming
- [x] Withdraw
  - [ ] Approve
  - [x] Confirm 
  - [x] Use selected account data in all steps
  - [ ] ~~Poll (regression test against account 0 on prod)~~ WS is super borked atm and Txs messages are not incoming
 

## Notice




## Pull Request Type


## Issue (if applicable)

- relates to the closed https://github.com/shapeshift/web/issues/2605
- relates to https://github.com/shapeshift/web/issues/2615
- relates to https://github.com/shapeshift/web/issues/2730

## Risk

Tested with MetaMask (which has only account 0 support) but this could bring regressions for the Yearn DeFi modal for single accounts, test accordingly

## Testing (with KeepKey only currently)

- Add an Ethereum account and send some ETH and an asset to stake to it to it
- Stake in some opportunity, reload the app (websocket is currently borked) and click on the opportunity card for it
- Select account 1 and click deposit
- Notice account #1 is pre-selected at Deposit step, with the right amount to deposit 
- Deposit some more of your staked asset, sign and broadcast it - it should be deposited from account #1
- Disconnect and reconnect wallet, then reopen the overview modal and select withdraw with account #1 selected in overview
- Notice account #1 is pre-selected at Withdraw step, with the right amount available to withdraw
- Withdraw some of your staked asset - it should be withdrawn from account #1
- Disconnect and reconnect wallet, then reopen the Cosmos overview modal with account #1 selected

### Engineering



- Refer to top-level testing steps
- Polling logic looks sound as it's not able to be tested currently
- Double-check the logic looks sound in deposit withdraw and overview for all steps

### Operations

- Refer to top-level testing steps



## Screenshots (non-exhaustive, high gas prices for those Txs with KK currently, so I didn't screenshot the confirm step, note that with current ws issue it will never show as completed in the UI, but you can confirm completion on Etherscan)

<img width="523" alt="image" src="https://user-images.githubusercontent.com/17035424/190500147-d6dad9fb-bca1-423a-a757-ae893b6c4932.png">
<img width="542" alt="image" src="https://user-images.githubusercontent.com/17035424/190500185-f99b72fe-a543-43eb-8d8b-77852111410e.png">
<img width="545" alt="image" src="https://user-images.githubusercontent.com/17035424/190501048-7d234a2d-0a13-4438-94a9-2fceae0ab5e5.png">
<img width="532" alt="image" src="https://user-images.githubusercontent.com/17035424/190500964-750a7e47-be60-43af-a983-4dfa300593b4.png">
<img width="525" alt="image" src="https://user-images.githubusercontent.com/17035424/190500235-7e0c15dc-0cdd-4806-9ada-3670a9d62371.png">
<img width="533" alt="image" src="https://user-images.githubusercontent.com/17035424/190501163-5d1ccd32-c327-4d4c-95ae-caec8ae252a9.png">
